### ISSUE TITLE
scrape_player_on_wall() does not scale well to high FPS

### ISSUE THREAD

**Issue by zicodxx** [dxx-rebirth/dxx-rebirth#310]
During testing today, I noticed that scrape_player_on_wall() as well as check_volatile_wall() in D2X do both not scale very well to FPS. In issue #276 we found out that this function can be called multiple times per frame causing excessive damage and I fixed this. And I can confirm that this fix does not seem to be a regression. The damage is fine. 

But the results from the contact with water or lava vary drastically depending on FPS.
1) Palette flashing is almost non existent at high FPS
2) random player bump effect caused by lava or water is much less significant at higher FPS 
3) the player can easily get stuck on such surfaces if they try to slide along (FPS > ~500)

Since these operations are executed in every Frame, this makes sense: the player bumping as well as the random rotvel changes are not scaled by FrameTime, yet executed in every frame. And since they are random in every frame, scaling them to FrameTime isn't the way to go - these pushes should be done in intervals. This is already done for the Fusion charging effect which works by the exact same rules - by using a d_tick_step check. But doing this, for some reason does not solve the issue. 

The player suddenly receives less damage than they should even if the check for d_tick_step just being applied to bump_one_object as well as the rotvel changes while apply_damage_to_player is still executed in every frame as it should.

What worries me the most is that PALETTE_FLASH_ADD should still properly add up as it takes the FrameTime-scaled 'damage' as argument and is applied once per frame, too. So this should work 'as is' already - but it doesn't. 

This should actually be easy to fix. As stated before, the Fusion Charge effect works identical. I feel I am missing a very obvious detail here. I will revise the issue tomorrow but if someone sees what I overlooked, please let me know. :)

**Comment by zicodxx** [comment]
Upon further investigation it seems that - no matter how high the FPS setting - scrape_player_on_wall is never called more than a maximum of 120 times per second. This explains why d_tick_step does not work here. And it also causes a bunch of other problems if this should apply to other functions called by do_physics_sim...

**Comment by zicodxx** [comment]
Investigating a bit more, I realize my mistake. Since each lava contact pushes the player away - with randomization in direction but constant strength - from the surface in every frame it makes sense that the per frame movement from the player isn't enough to counter this at a specific Framerate. 
Ergo, scaling the damage to FrameTime is the wrong approach since the contact to the damaging surface (or any surface for that matter, depending on the sliding angle) is not intended to happen over a constant period of time - it's only 'possible' to happen at lower Framerates.
Instead, the shaking, pushing as well as the damage infliction - all three - should be done 'per contact' and regulated in intervals. I.e. like the Fusion charge effect but with the difference that we can't utilize d_tick_step as this isn't guaranteed to be true when a player makes contact with the surface. It needs its own timer for that.

**Comment by zicodxx** [comment]
I pushed b36c6f20c705d5eae78f349794a337612c9c8d14 which fixes the issue.

I did write an explanation in the commit message but I can go in a little bit more detail here:
Generally it would make sense to scale damage based on how long the player is in contact with a damaging surface. However this damage causes the player to have random changes in rotvel in every single frame - which in itself is bad enough as it is and at the same time it causes bump_one_object in the same fashion as well. 

Since both happen per frame while not properly scaled to FrameTime (which it can't since there is a new random vector per frame), the player movement could not counter this which made it eventually impossible for the player object to maintain contact with the surface in ever frame. 

All in all, the way how scrape_player_on_wall (and check_volatile_wall in D2) is (are) working, there is a logical misconception since damage is supposed to be scaled per frame so the player can safely maintain contact with the surface while at the same time you get pushed away from it, breaking up this continuous contact. Of course the problem only manifests at FPS which are considerably higher than possible when the game was released but it's still a function which logic undermines its own purpose.

Instead I am using a timer which technically only allows contact with the surface 30 times per second and scales the damage output accordingly. At the same time this solves the issue that the random changes rotvel and player bumps do not unintentionally increment with higher FPS.

I am of course not happy to have yet another timer in this function (the first one making sure the sound effects only play 4 times per second) but considering how we would still need a timer in a per frame approach to make to mitigate new random vectors in every frame, I think this is the best solution for how this function currently works. 

It goes without saying that this approach only works for a single object but since it is only meant to apply to the local player object, this is fine.
